Enable STM32L4R9i-disco DSI display via shield ST_B_DSI_MB1314

avolmat-st opened the pull request:
This pull request includes updates, new driver and a new shield in order to enable the DSI display that can be mounted on the STM32L4R9I-DISCO board.
It enables both LTDC and DSI interfaces of the STM32L4R9 and for that purpose require the usage of the PLLSAI.

PLLSAI support isn't yet merged. There is another ongoing pull request for that purpose which can be found here:
https://github.com/zephyrproject-rtos/zephyr/pull/85135
The PLLSAI pull-request is still evolving hence this new pull-request will be rebased afterward once the PLLSAI will be merged.

Since it rely on another pull-request, it won't build (missing pllsai nodes or macros) but I am already pushing this one in order to start the review process.

Commit:
boards: shields: add st_b_dsi_mb1314 shield (DSI display)

Addition of a new shield which embeds a round DSI display
gvo,g1120tb101 of 390x390 pixels, with a touchpanel.
This shield enables it via usage of the LTDC and DSI
of the STM32.

Touchpanel isn't yet supported in this commit and will
be added later on in an additional commit.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
display: addition of GVO G1120TB101 DSI panel

Add support for the GVO G1120TB101 DSI Panel.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
dts: st: l4: stm32l4r9: addition of mipi_dsi node

stm32l4r9 series embeds a MIPI DSI controller which uses the
LTDC controller as input.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
dt-bindings: clock: stm32l4: addition of HSE clock source

Addition of the STM32_SRC_HSE define in stm32l4_clock.h in
order to be able to address the HSE clock source from the
device-tree.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
dts: arm: stm32l4: add ltdc in stm32l4p5 and stm32l4q5

ltdc controller is also available in stm32l4r5 and stm32l4q5
so add it within stm32l4p5.  It is not available in stm32l4r5
hence delete the node for this platform.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
mipi_dsi: stm32: use mdev pixfmt to configure DSI interface

Rely on the pixel format defined within the display node
in order to configure the DSI interface pixel format.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
display: stm32_ltdc: use DISPLAY_BITS_PER_PIXEL to compute format BPP

Use the macro DISPLAY_BITS_PER_PIXEL in order to compute each format
BPP and thus avoid requiring each format BPP written within the
driver.
Thanks to that, simply the function stm32_ltdc_set_pixel_format to
avoid code duplication.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

mathieuchopstm commented on drivers/display/display_g1120tb101.c:
```diff
@@ -0,0 +1,273 @@
+/*
+ * Copyright (c) 2025 ST Microelectronics
+ * Inpired from display_otm8009a.c
```
Typo:
```suggestion
 * Inspired from display_otm8009a.c
```

mathieuchopstm commented on drivers/display/display_stm32_ltdc.c:
```diff
@@ -45,20 +45,19 @@ LOG_MODULE_REGISTER(display_stm32_ltdc, CONFIG_DISPLAY_LOG_LEVEL);
 #else
 #error "Invalid LTDC pixel format chosen"
 #endif
+#define STM32_LTDC_INIT_PIXEL_SIZE	DISPLAY_BITS_PER_PIXEL(DISPLAY_INIT_PIXEL_FORMAT) /\
+					BITS_PER_BYTE
```
I think this *might* be more readable if the `/` is on the second line?

avolmat-st commented on drivers/display/display_stm32_ltdc.c:
```diff
@@ -107,25 +106,23 @@ static int stm32_ltdc_set_pixel_format(const struct device *dev,
+	} else if (format == PIXEL_FORMAT_ARGB_8888) {
+		ltdc_pix_fmt = LTDC_PIXEL_FORMAT_ARGB8888;
+	} else {
+		return -ENOTSUP;
```
original code didn't have an ERR_LOG for non supported format so I propose to keep it like that in this PR.

Commit:
boards: st: add DSI connector and display alias in stm32l4r9i_disco

Add the description of the DSI connector available on
the stm32l4r9i_disco board and zephyr_mipi_dsi/zephyr_lcd_controller
alias to be used by display shields.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

avolmat-st commented:
Triggered again the test to allow proper building since now PLLSAI support has been merged.
Tested ok merged on latest main branch using

west build -p -b stm32l4r9i_disco --shield st_b_dsi_mb1314 samples/drivers/display/

avolmat-st commented on boards/shields/st_b_dsi_mb1314/boards/stm32l4r9i_disco.overlay:
```diff
@@ -0,0 +1,49 @@
+/ {
+	chosen {
+		/* SRAM0 is too small to hold a framebuffer */
+		zephyr,sram = &sram2;
```
Good idea. Let me work on that, hopefully might be able to use the zephyr,memory-region / zephyr,memory-label in order to generate the linker entry for all regions and then later on be able to point to any of it.
That would allow to remove the current code allowing to point simple to several dedicated sdram1/sdram2/psram.

avolmat-st replied:
This is actually possible but would like to do that correctly for all STM32 board which rely on the ext-sdram and linker scripts in memc so will first do that in a separate PR in order to be clean and then update here again this PR once the modification is done.

Ex of what I did:

```
+#define FRAME_BUFFER_SECTION   \
+       Z_GENERIC_SECTION(LINKER_DT_NODE_REGION_NAME(DT_INST_PHANDLE(0, ext_sdram)))
```

Assuming the phandle given in ext_sdram has the right compatible and property, this will allow to get the section attribute automatically, and thus remove the need to have dedicated cases for all memory:

```
+               compatible = "zephyr,memory-region", "mmio-sram";
                reg = <0x10000000 DT_SIZE_K(64)>;
+               zephyr,memory-region = "SRAM1";
```

avolmat-st replied:
Base implemented within PR #95299

Commit:
tests: display: display_check: add st_b_dsi_mb1314 shield

Addition of tests with the st_b_dsi_mb1314 shield.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

avolmat-st commented:
Twister failure is due to the need to have this PR applied on top of the PR #95299.

avolmat-st commented on drivers/display/display_g1120tb101.c:
```diff
@@ -0,0 +1,273 @@
+{
+	uint8_t InitParam1[4] = {0x00, 0x04, 0x01, 0x89};
+	uint8_t InitParam2[4] = {0x00, 0x00, 0x01, 0x85};
+	int ret, i;
```
Is your proposal related to MISRA rule about multiple declaration within the same line ?
(A declaration should not declare more than one variable or member variable)

I am not sure this is considered as a requirement in Zephyr CodyingStyle and doesn't seem to be listed in required rule right ? (doc/contribute/coding_guidelines/index.rst)

avolmat-st commented on drivers/display/Kconfig.g1120tb101:
```diff
@@ -0,0 +1,20 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config G1120TB101
+	bool "G1120TB101 display controller"
```
Ouch ... THANKS for spotting this.
I also checked again, I originally based this on the Cube Example but mixed up the panel / controller.
This changes things since the RM67162 is actually already supported with a driver in Zephyr so I will rework this part of the PR.
